### ISSUE TITLE
extra_toolchains precedence not respected when duplicates occur

### ISSUE THREAD

**Issue by rickeylev** [bazelbuild/bazel#22912]
### Description of the bug:

Per https://github.com/bazelbuild/bazel/issues/19945, it was decided that elements later in the `--extra_toolchains` list should have higher precedence. This mostly works, but seems to regress when there are duplicates of a target in the list. When that happens, the _first_ position of a target takes precedence instead of the last.

This isn't too big a deal, but does makes having a `--extra_toolchains` value with consistent behavior between Bazel 6 and Bazel 7 a bit of a pain. Specifically, there isn't a single list that can get the desired behavior in both versions. For example, lets say we have two toolchain targets, `//a:all` and `//b:all`. We want to always use `b`.

```
extra_toolchains=[a, b]
Bazel 6: picks a
Bazel 7: picks b

extra_toolchains=[b, a]
Bazel 6: picks b
Bazel 7: picks a

extra_toolchains=[b, a, b]
Bazel 6: picks b
Bazel 7: picks a  <-- err wait what?

# Hm, lets repeat the same value in different ways...
extra_toolchains=[a, b, a]
Bazel 6: picks a
Bazel 7: picks b

extra_toolchains=[a, b, a, b]
Bazel 6: picks a
Bazel 7: picks b
```

So yeah, it seems like the first position in the list determines the final position when iterating. This smells very much like LinkedHashSet (first added is iteration order), or depset (first iterated upon is yielded, subsequents are ignored) doing something, somewhere.

I skimmed through the RegisterToolchainsFunction.java code, but nothing in particular stuck out to me.

cc @katre 

### Which category does this issue belong to?

Configurability

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Pass the same toolchain twice using `--extra_toolchains`. Bazel 7 won't respect the "last has priority" as described in https://github.com/bazelbuild/bazel/issues/19945

I can put together a repro tonight.

### Which operating system are you running Bazel on?

linux

### What is the output of `bazel info release`?

7.2.1

### If this is a regression, please try to identify the Bazel commit where the bug was introduced with bazelisk --bisect.

Not quite a regression, but https://github.com/bazelbuild/bazel/commit/75f8017ba3a87da4f593e37f0eb8f333b330d95c from https://github.com/bazelbuild/bazel/issues/19945 is related

### Have you found anything relevant by searching the web?

https://github.com/bazelbuild/bazel/issues/19945

**Comment by katre** [comment]
Ha ha sigh.

The problem here is in the normalization used (see [`PlatformOptions.getNormalized`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/PlatformOptions.java;drc=74c8ca4886cfc188d60a18e796743e842f675842;l=180)): it calls into [`dedupeOnly`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/config/FragmentOptions.java?q=symbol%3A%5Cbcom.google.devtools.build.lib.analysis.config.FragmentOptions.dedupeOnly%5Cb%20case%3Ayes), which keeps the **first** instance of the value, and so shows the behavior you report.

This flag (and maybe also `--extra_execution_platforms`?) really wants a currently-doesn't-exist `dedupeOnlyKeepingLast`, instead.

This probably isn't difficult, I'll take a look at it sometime soon.

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.4.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.4.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=7.4.0rc1`. Thanks!
